### DISCUSSION TITLE
How to generate a sequence of rules that produced the AST?

### DISCUSSION THREAD

**Discussion by brando90** [lark-parser/lark#918] [Q&A] [answered]
Hi,

I was wondering, during the parsing of the string one creates an AST. I was wondering if there was an easy way to extract the rules that generated a specific AST from a specific order traversal ? (note: non-terminals be expanded in different orders e.g. BFS or DFS so the sequence would correspond to such an ordering)?

**Comment by MegaIng** [comment, marked as answer]
There is the [`Transformer`,  `Visitor` and `Interpreter`](https://lark-parser.readthedocs.io/en/latest/classes.html#transformer-visitor-interpreter) classes that visit the nodes of the tree.

If that does not do what you want, you will need to provide an example of what you want. Potentially the internal [`tree_matcher`](https://github.com/lark-parser/lark/blob/master/lark/tree_matcher.py) does what you want.

Note that if you want to match how it is parsed in reverse, that doesn't really make sense since we are implementing different algorithms that don't directly walk the grammar in a tangible way.

**Reply by brando90** [reply]
I will take a look at that. Meanwhile, let me try one more time describe what I want while I think of a concrete example + go through the links you sent me.

What I want is to generate a sequence of rules (say in BFS ordering) that if I visited the non-terminals in the order the rule sequence was generated, then I would generate the given string in the language. So from a string I generate a sequence of rules that would generate that string if I were to execute the rules in the order that was used to generate the sequence.

Let me know if that makes more sense.

Thanks for the pointers - will check them out and figure out a small example.

**Reply by MegaIng** [reply]
If your goal is to 'reconstruct' the source string, the [`Reconstructor`](https://lark-parser.readthedocs.io/en/latest/examples/advanced/reconstruct_json.html) does exactly that. The [source code](https://github.com/lark-parser/lark/blob/master/lark/reconstruct.py) also is the primary use for the tree_matcher, which means you can probably use tree_matcher it to do what you want here, if the Reconstructor isn't enough.

**Reply by brando90** [reply]
My goal is not to reconstruct the source string. My goal is to have a sequence of grammar rules that traversed in a pre-define manner generates the string (perhaps more context will make this sound less random to you, what I want to predict this sequence of grammar rules using a neural network/machine learning). The actual generation of the string is a side benefit but what I want is this API:
```
term_in_grammar -> sequence_of_grammar_rules_for_that_term
```
e.g.
```
1+1 -> [rule1, rule2, ..., rule_k] (ordered by BFS)
```
Perhaps that's more clear. 

If could label the edges of the AST with the rule that generate the non-terminal and terminal would be enough I believe.

**Reply by brando90** [reply]
To make my example more concrete, let's consider this grammar:
```
exp ::= term | term + term
term := num | var
num := 1 | 2
var := x | y
```
then 1 becomes:
```
1 -> [rule1, rule3, rule5]
```
since there is no AST DFS and BFS give the same ordering.

**Reply by erezsh** [reply]
I don't really understand the difficulty.

Given an input string, lark returns the rules that match it, in the structure in which they are nested.

You can traverse that tree any way you like, and extract the rule names in whatever order suits you.

What do you need that isn't covered by that?

**Reply by brando90** [reply]
btw, I still have in my list of TODOs to do all your suggestions! I think I am nearly there at doing that. I appreciate your feedback and helping me clarify my thinking! :) This was useful progress to me and now I know what things I need to read in Lark.

**Reply by brando90** [reply]
> I don't really understand the difficulty.
> 
> Given an input string, lark returns the rules that match it, in the structure in which they are nested.
> 
> You can traverse that tree any way you like, and extract the rule names in whatever order suits you.
> 
> What do you need that isn't covered by that?

We just intersected in comments. Oh I think it's just going through the suggested things! Thanks! If I get an issue I will comment.

Appreciate the help/feedback!

**Reply by brando90** [reply]
> If your goal is to 'reconstruct' the source string, the [`Reconstructor`](https://lark-parser.readthedocs.io/en/latest/examples/advanced/reconstruct_json.html) does exactly that. The [source code](https://github.com/lark-parser/lark/blob/master/lark/reconstruct.py) also is the primary use for the tree_matcher, which means you can probably use tree_matcher it to do what you want here, if the Reconstructor isn't enough.

@MegaIng The issue with `reconstruction` is that it assumes I already have a parse tree. I do not have a parse tree or the text - instead I am trying to generate that parse tree from a list of rewrite rules (note a re-write rule is `alpha -> beta` where `alpha, beta` are symbols for collections of non-terminals and terminals).

**Reply by MegaIng** [reply]
@brando90 Assuming you have a magic function that does exactly what you want it to right now: What is the input, what is the output as concrete examples? I am having a hard time understanding what you want to do.

**Reply by brando90** [reply]
> Given an input string, lark returns the rules that match it, in the structure in which they are nested.
> 
> You can traverse that tree any way you like, and extract the rule names in whatever order suits you.
> 
> What do you need that isn't covered by that?

Hi @ealter! Thanks for the patience to reply. The issue is the assumption that I start off from a string or a parse tree. In fact I start of from a sequence of re-write rules `alpha->beta` (where `alpha, beta` are any symbol - terminals or non-terminals). From this sequence of re-write rules I am trying to construct a parse tree. Note that the sequence of rules could in theory not generate a valid parse tree but that would be easy to detect in a context-free grammar since it's "markovian" (i.e. at some point the re-write rule would not apply correct because something does not match).

**Reply by MegaIng** [reply]
@brando90 Aha. You mean rules in the sense of [Formal Grammar](https://en.wikipedia.org/wiki/Formal_grammar) [Production Rules](https://en.wikipedia.org/wiki/Production_(computer_science))?

**Reply by brando90** [reply]
> @brando90 Assuming you have a magic function that does exactly what you want it to right now: What is the input, what is the output as concrete examples? I am having a hard time understanding what you want to do.

Assume this:

```
grammar = r" 
S : "a" \\ rule 1
  | "a" S \\ rule 2
"

# rule sequence according to BFS that corresponds to the string "aa"
rule_seq = ['S -> "a" S', 'S -> "a"' ]
rule_seq1 = [1,  2]
ast = rule_seq2lark_tree(rule_seq)
text = reconstruction(ast) # has string "aa"
```

The above is a small example of the entire flow. Note that I get `rule_seq` randomly - not necessarily from strings themselves.

**Reply by brando90** [reply]
Perhaps it's not easy to use lark or its functionality to do it and I have to do it from scratch - and that's alright too. I can figure it out but wanted to check if this was already easy with Lark.

**Reply by MegaIng** [reply]
If you end goal is to generate random strings from lark grammars, there is a tool for that: [hypothesis.extra.lark](https://hypothesis.readthedocs.io/en/latest/_modules/hypothesis/extra/lark.html) (But it is far from optimal).

But to your question: No, lark does not currently have a function that does that. But it should not be to impossible to build. A `Lark` instance has a `.rules` property. I will see what I can do and report back to you.

**Reply by MegaIng** [reply]
@brando90 Here a small implementation as a starting point:

```
from collections import defaultdict

from lark import Lark, Tree, Token
from lark.grammar import Rule, NonTerminal, Symbol, Terminal
from lark.reconstruct import Reconstructor


def build(rules: list[Rule], rule_seq: list[int], build_term) -> Tree:
    def build_rule(rule: Rule) -> Tree:
        l = []
        for i, e in enumerate(rule.expansion):
            if e.is_term:
                l.append(build_term(e))
            else:
                l.append(e)
                targets[e].append((l, i))
        return Tree(rule.origin.name, l)

    out: list = [NonTerminal("start")]
    targets = defaultdict(list)
    targets[out[0]].append((out, 0))
    for i in rule_seq:
        r = rules[i]
        assert r.alias is None, "Can't have aliases"
        assert r.options.keep_all_tokens, "need to have keep_all_tokens"
        assert not (r.options.expand1 or r.origin.name.startswith("_")), "Can't have a rule that expands it's children"
        ts = targets[r.origin]
        l, i = ts.pop(0)
        assert l[i] == r.origin, l
        l[i] = build_rule(r)
    return out[0]


grammar = r"""
start: "a" // rule 0
     | "a" start // rule 1
"""

parser = Lark(grammar, keep_all_tokens=True)

print(parser.rules)

rule_seq1 = [1, 0]
ast = build(parser.rules, rule_seq1, lambda t: Token(t.name, "a"))
print(ast)
text = Reconstructor(parser).reconstruct(ast, None)  # has string "aa"
print(repr(text))
```

Currently somewhat limited, but it shouldn't be to hard to extend it to include everything you need.

**Reply by brando90** [reply]
@MegaIng thanks for that! I had something I think that is similar but I was hard coding the rules tuples and other ugly things. It's nice that the parser gives them for me with `.rules`! Appreciate it!

**Reply by MegaIng** [reply]
@brando90 I expanded it a bit: https://github.com/MegaIng/lark_ast_generator

@erezsh It might be worth adding a tool like this to lark itself somewhere. This is not the first time someone asked for something like this.

**Reply by ThatXliner** [reply]
I keep on getting notifications for this (I'm watching the repo), and I'm not interested until I saw a link to a repo.

@brando90, what did you want, exactly?

**Reply by brando90** [reply]
@MegaIng Thanks for that! 

I was also interested in generating `rule_rewrite_sequences` from ASTs. The algorithm at a high level is simple to generate rule sequences: 

> from each (parent) node in a AST, get it's children to form the rule in the form `Rule( parent -> children1...children_N)` and then go through the `grammars.rules` to see which one it matches to get the index.

I am wondering if your code named:

https://github.com/MegaIng/lark_ast_generator/blob/a0c09596e98c40746c98a9aad2b73b54026e267d/ast_generator.py#L114

already does that or if I need to generate it from scratch traversing the AST?

In summary: just checking if you already know how to do `rule_seq = generate_rule_sequence_from(AST)` in Lark since you are a Lark expert rather than me hacking away something ugly.

Thanks!

PS: btw, I've been doing everything with a BFS ordering, but I think my implementation can be changed easily by changing the queue into a stack.

**Reply by brando90** [reply]
> I keep on getting notifications for this (I'm watching the repo), and I'm not interested until I saw a link to a repo.
> 
> @brando90, what did you want, exactly?

Hi @ThatXliner! I am trying to transform a string to an AST to a sequence of rewrite-rules that if executed in some ordering yields that string (and the reverse too - from the string to the AST to the sequence of rewrite-rules to the AST to the string).

The crucial component that is not the "standard Lark" thing I believe is that the **sequence of rewrite-rules** are the representation of the term instead of an AST (now I am realizing that). What I have as input are the rule sequences that need to be generated or outputted transformed to ASTS (back and forward).

**Reply by ThatXliner** [reply]
That's a lot of words. I'm visualizing

`string -> ast -> rules -> ast -> string`

So you want to extract the `rules` from `ast`, correct? Guaranteed no matter what ordered it was generated?

**Reply by brando90** [reply]
> string -> ast -> rules -> ast -> string

Yes! Right now I have `seq_rules->ast->string`. Now I am missing `string->ast->seq_rules` (but I outlined an algorithm that would work though I am hoping there is a Lark magic that makes that easier or I will just implement my suggestion).

Hope that clarifies it. Apologies on the wordiness. You are right I need to work on that!

Your question on the "order" not sure what you mean by the "no matter what order". The order certainly matters because the algorithm has to assume the sequence of rewrite rules are given in some order (e.g. dfs, bfs) otherwise it cannot possible generate a AST from a sequence (the treeness info is lost from a sequence so that has to be assumed or specified somehow or guessed). I don't think it's worth trying to write a "general" one that accepts any order (e.g. like given a lambda function that accepts any order...even random transversals, I'm thinking very general otherwise it will be very hard to index the queue). I just hardcoded bfs by poping left from a dequeue. I think it's better to have code that if the sequence doesn't respect the tree traversal that it throws an error.  Again lot's of words but I hope that somewhat clarifiers.

**Reply by MegaIng** [reply]
@brando90 `tree_matcher` might be able to do AST->rules. I will see what I can do.

**Reply by brando90** [reply]
@MegaIng got something sort of working...but I just realized that the method I described in English a few posts ago assumes that we only have terminals and non-terminals. For example, if we have a long list of children that was generated from a `*` or a `+` then I get can't just get all the children into the expansion and call the parent the origin and call that a rule. Will think about it more or look into tree_matcher...

**Reply by MegaIng** [reply]
@brando90 I mean, yeah, none of these algorithms work on EBNF, but only on BNF. I think I got something working, I will update the repo soon.

**Reply by brando90** [reply]
I guess if I could transform my grammar to a format without stars etc that would work e.g.

```
A : "a" A+ ==> A :  "a" A
A : "a" A* ==> A: "a" | "a" A
A : "a" [A] ==> A: "a" | "a" "a"
```
ok, this works for this grammar but it seems rather subtle. Let me know if you have a better solution using Lark magic. Will think more.

**Reply by MegaIng** [reply]
@brando90 Yes. lark does this under the hood, e.g. `parser.rules` is in BNF format.

**Reply by brando90** [reply]
> @brando90 Yes. lark does this under the hood, e.g. `parser.rules` is in BNF format.

oh ok. I will wait for your reply then. Thanks for your time and patience! :)

**Reply by MegaIng** [reply]
@brando90 Ok, I just pushed the [rules_generator](https://github.com/MegaIng/lark_ast_generator/blob/master/rules_generator.py). You can see how to use it from [test.py](https://github.com/MegaIng/lark_ast_generator/blob/master/test.py). It is not perfect, and will as example introduce unnecessary parentheses. (But that is also partly because of a bug I found in tree_matcher I think).

**Reply by brando90** [reply]
@MegaIng awesome! Will go through it now. The attempt I described worked for simple grammars (bnf) but it does work at all with the ebnf one that Lark supports. If I see but fixes I will let you know.

**Reply by brando90** [reply]
@MegaIng Hi Megalng, I was going through the code and was finding that it is trickier to understand that I expected (probably my own inexperience with Lark than anything else + not knowing even conceptually what the algorithm should be for ebnfs) but was wondering if you'd mind explaining the main idea(s) in English? I was hoping that would help me understand what I am reading (since right now I am not even sure what the intended algorithm is and trying to infer that only from source code is sort of tricky - perhaps due to implementation details).

Thanks again for getting that together for me! And your time is very much appreciated.

**Reply by MegaIng** [reply]
@brando90 I will go through it and create comments. For further questions you can just open an issue over there. Or maybe even better, write me on gitter.

**Reply by MegaIng** [reply]
@brando90 I pushed a bit of documentation in rules_generator. For further questions, I would suggest opening an [issue](https://github.com/MegaIng/lark_ast_generator/issues/new) over at the repo or sending me a PM on [gitter](https://gitter.im/lark-parser/Lobby).
